### PR TITLE
Replacing IExpirationTrigger with IChangeToken

### PR DESCRIPTION
**Pull request by pranavkm**

### CODE REVIEW

File: src/Microsoft.Framework.Caching.Abstractions/CancellationTokenTrigger.cs [hunk 1]
```diff
@@ -3,31 +3,33 @@
...
 using System.Threading;
+using Microsoft.Framework.Primitives;
 
 namespace Microsoft.Framework.Caching
 {
-    public class CancellationTokenTrigger : IExpirationTrigger
+    /// <summary>
+    /// A <see cref="IChangeToken"/> implementation using <see cref="CancellationToken"/>.
+    /// </summary>
+    public class CancellationTokenTrigger : IChangeToken
```
Comment (pranavkm) [inline src/Microsoft.Framework.Caching.Abstractions/CancellationTokenTrigger.cs]: 1. Should we rename this? `CancellationChangeToken` \ `CancellationTokenBasedChangeToken`?
2. FileProvider uses this type. Should we move this to Common or have FileSystem replicate it?

File: src/Microsoft.Framework.Caching.Abstractions/MemoryCacheEntryOptions.cs [hunk 2]
```diff
@@ -76,7 +77,7 @@ public class MemoryCacheEntryOptions
         /// <summary>
         /// Gets or sets the events which are fired when the cache entry expires.
         /// </summary>
-        public IList<IExpirationTrigger> Triggers { get; } = new List<IExpirationTrigger>();
+        public IList<IChangeToken> Triggers { get; } = new List<IChangeToken>();
```
Comment (pranavkm) [inline src/Microsoft.Framework.Caching.Abstractions/MemoryCacheEntryOptions.cs]: I'm guessing we don't want to rename `Triggers` the property name.

Comment (Tratcher) [inline src/Microsoft.Framework.Caching.Abstractions/MemoryCacheEntryOptions.cs]: It doesn't line up anymore, open to suggestions.

File: src/Microsoft.Framework.Caching.Abstractions/CancellationTokenTrigger.cs (hunk 1 above)
Comment (Tratcher) [inline src/Microsoft.Framework.Caching.Abstractions/CancellationTokenTrigger.cs]: It's so dirt simple, move it to common.

Comment (Tratcher) [inline src/Microsoft.Framework.Caching.Abstractions/CancellationTokenTrigger.cs]: CancellationTokenToken... :-)

File: src/Microsoft.Framework.Caching.Abstractions/MemoryCacheEntryOptions.cs (hunk 2 above)
Comment (Tratcher) [inline src/Microsoft.Framework.Caching.Abstractions/MemoryCacheEntryOptions.cs]: Triggers should be replaced in the whole solution, it doesn't make sense anymore. AddExpirationTrigger(IChangeToken)...

**Comment by pranavkm** [comment]
Updated

File: src/Microsoft.Framework.Caching.Abstractions/MemoryCacheEntryExtensions.cs [hunk 3]
```diff
@@ -24,17 +25,17 @@ public static class MemoryCacheEntryExtensions
         /// Expire the cache entry if the given event occurs.
         /// </summary>
         /// <param name="options"></param>
-        /// <param name="trigger"></param>
-        public static MemoryCacheEntryOptions AddExpirationTrigger(
+        /// <param name="token"></param>
+        public static MemoryCacheEntryOptions AddExpirationToken(
```
Comment (Tratcher) [inline src/Microsoft.Framework.Caching.Abstractions/MemoryCacheEntryExtensions.cs]: AddChangeToken?

File: src/Microsoft.Framework.Caching.Memory/CacheEntry.cs [hunk 4]
```diff
@@ -82,17 +82,17 @@ private bool CheckForExpiredTime(DateTimeOffset now)
             return false;
         }
 
-        internal bool CheckForExpiredTriggers()
+        internal bool CheckForExpiredTokens()
```
Comment (Tratcher) [inline src/Microsoft.Framework.Caching.Memory/CacheEntry.cs]: ChangedTokens?

**Comment by Tratcher** [comment]
Am I the only one that thinks these name changes sound very out of place here? @muratg @Eilon

File: src/Microsoft.Framework.Caching.Abstractions/EvictionReason.cs [hunk 5]
```diff
@@ -25,7 +25,7 @@ public enum EvictionReason
         /// <summary>
         /// Event
         /// </summary>
-        Triggered,
+        TokenChanged,
```
Comment (Eilon) [inline src/Microsoft.Framework.Caching.Abstractions/EvictionReason.cs]: Maybe `TokenExpired`?

File: src/Microsoft.Framework.Caching.Abstractions/IEntryLink.cs [hunk 6]
```diff
@@ -19,15 +20,15 @@ public interface IEntryLink : IDisposable
         DateTimeOffset? AbsoluteExpiration { get; }
 
         /// <summary>
-        /// Gets all the triggers from the dependent entries.
+        /// Gets all the <see cref="IChangeToken"/>s from the dependent entries.
         /// </summary>
-        IEnumerable<IExpirationTrigger> Triggers { get; }
+        IEnumerable<IChangeToken> ChangeTokens { get; }
```
Comment (Eilon) [inline src/Microsoft.Framework.Caching.Abstractions/IEntryLink.cs]: ExpirationTokens?

File: src/Microsoft.Framework.Caching.Abstractions/IEntryLink.cs [hunk 7]
```diff
@@ -19,15 +20,15 @@ public interface IEntryLink : IDisposable
...
 
         /// <summary>
-        /// Adds triggers from a dependent entries.
+        /// Adds <see cref="IChangeToken"/>s from a dependent entries.
         /// </summary>
-        /// <param name="triggers"></param>
-        void AddExpirationTriggers(IList<IExpirationTrigger> triggers);
+        /// <param name="changeTokens"><see cref="IChangeToken"/>s from dependent entries.</param>
+        void AddChangeTokens(IList<IChangeToken> changeTokens);
```
Comment (Eilon) [inline src/Microsoft.Framework.Caching.Abstractions/IEntryLink.cs]: AddExpirationTokens? And expirationTokens

File: src/Microsoft.Framework.Caching.Abstractions/MemoryCacheEntryExtensions.cs [hunk 8]
```diff
@@ -24,17 +25,17 @@ public static class MemoryCacheEntryExtensions
...
             this MemoryCacheEntryOptions options,
-            IExpirationTrigger trigger)
+            IChangeToken token)
```
Comment (Eilon) [inline src/Microsoft.Framework.Caching.Abstractions/MemoryCacheEntryExtensions.cs]: `expirationToken`

File: src/Microsoft.Framework.Caching.Abstractions/MemoryCacheEntryOptions.cs [hunk 9]
```diff
@@ -74,9 +75,9 @@ public class MemoryCacheEntryOptions
         }
 
         /// <summary>
-        /// Gets or sets the events which are fired when the cache entry expires.
+        /// Gets or sets the <see cref="IChangeToken"/>s which are notified when the cache entry expires.
         /// </summary>
-        public IList<IExpirationTrigger> Triggers { get; } = new List<IExpirationTrigger>();
+        public IList<IChangeToken> ChangeTokens { get; } = new List<IChangeToken>();
```
Comment (Eilon) [inline src/Microsoft.Framework.Caching.Abstractions/MemoryCacheEntryOptions.cs]: `ExpirationTokens`

File: src/Microsoft.Framework.Caching.Memory/CacheEntry.cs [hunk 10]
```diff
@@ -43,7 +43,7 @@ internal class CacheEntry
 
         internal EvictionReason EvictionReason { get; private set; }
 
-        internal IList<IDisposable> TriggerRegistrations { get; set; }
+        internal IList<IDisposable> ChangeTokenRegistrations { get; set; }
```
Comment (Eilon) [inline src/Microsoft.Framework.Caching.Memory/CacheEntry.cs]: ExpirationTokenRegistrations

**Comment by muratg** [comment]
:watch:

**Pull request closed without merging**